### PR TITLE
Fixed-point functions can take specific precision

### PR DESCRIPTION
**Pull request by Rangi42**
This allows concise single expressions to differ from a project's usual fixed-point precision. It goes well with the "specific precision" `q` literals. For example:

```
pusho
opt Q8
db mul(6.0, 7.0)
popo
```

can become:

```
db mul(6.0q8, 7.0q8, 8)
```

### CODE REVIEW

File: src/asm/fixpoint.c [hunk 1]
```diff
@@ -21,105 +20,96 @@
 #define M_PI 3.14159265358979323846
 #endif
 
-#define fix2double(i)	((double)((i) / fix_PrecisionFactor()))
-#define double2fix(d)	((int32_t)round((d) * fix_PrecisionFactor()))
+#define fix2double(i)	((double)((i) / pow(2.0, q)))
+#define double2fix(d)	((int32_t)round((d) * pow(2.0, q)))
```
Comment (ISSOtm) [inline src/asm/fixpoint.c:24]:
```suggestion
#define fix2double(i, q)	((double)((i) / pow(2.0, q)))
#define double2fix(d, q)	((int32_t)round((d) * pow(2.0, q)))
```

I was very confused by where the `q` argument was used. I understand why you did this, but I don't think it's a good idea.

File: src/asm/fixpoint.c [hunk 2]
```diff
@@ -21,105 +20,96 @@
...
-
-	printf("%s%" PRIu32 ".%05" PRIu32, sign, u >> fixPrecision,
-	       ((uint32_t)(fix2double(u) * 100000 + 0.5)) % 100000);
+	return pow(2.0, fixPrecision);
 }
 
-int32_t fix_Sin(int32_t i)
+int32_t fix_Sin(int32_t i, int32_t q)
 {
 	return double2fix(sin(turn2rad(fix2double(i))));
```
Comment (ISSOtm) [inline src/asm/fixpoint.c:44]:
```suggestion
	return double2fix(sin(turn2rad(fix2double(i, q))), q);
```

File: src/asm/fixpoint.c [hunk 3]
```diff
@@ -21,105 +20,96 @@
...
 }
 
-int32_t fix_Cos(int32_t i)
+int32_t fix_Cos(int32_t i, int32_t q)
 {
 	return double2fix(cos(turn2rad(fix2double(i))));
```
Comment (ISSOtm) [inline src/asm/fixpoint.c:49]:
```suggestion
	return double2fix(cos(turn2rad(fix2double(i, q))),q );
```

File: src/asm/fixpoint.c [hunk 4]
```diff
@@ -21,105 +20,96 @@
...
 }
 
-int32_t fix_Tan(int32_t i)
+int32_t fix_Tan(int32_t i, int32_t q)
 {
 	return double2fix(tan(turn2rad(fix2double(i))));
```
Comment (ISSOtm) [inline src/asm/fixpoint.c:54]:
```suggestion
	return double2fix(tan(turn2rad(fix2double(i, q))), q);
```

File: src/asm/fixpoint.c [hunk 5]
```diff
@@ -21,105 +20,96 @@
...
 }
 
-int32_t fix_ASin(int32_t i)
+int32_t fix_ASin(int32_t i, int32_t q)
 {
 	return double2fix(rad2turn(asin(fix2double(i))));
```
Comment (ISSOtm) [inline src/asm/fixpoint.c:59]:
```suggestion
	return double2fix(rad2turn(asin(fix2double(i, q))), q);
```

File: src/asm/fixpoint.c [hunk 6]
```diff
@@ -21,105 +20,96 @@
...
 }
 
-int32_t fix_ACos(int32_t i)
+int32_t fix_ACos(int32_t i, int32_t q)
 {
 	return double2fix(rad2turn(acos(fix2double(i))));
```
Comment (ISSOtm) [inline src/asm/fixpoint.c:64]:
```suggestion
	return double2fix(rad2turn(acos(fix2double(i, q))), q);
```

File: src/asm/fixpoint.c [hunk 7]
```diff
@@ -21,105 +20,96 @@
...
 }
 
-int32_t fix_ATan(int32_t i)
+int32_t fix_ATan(int32_t i, int32_t q)
 {
 	return double2fix(rad2turn(atan(fix2double(i))));
```
Comment (ISSOtm) [inline src/asm/fixpoint.c:69]:
```suggestion
	return double2fix(rad2turn(atan(fix2double(i, q))), q);
```

File: src/asm/fixpoint.c [hunk 8]
```diff
@@ -21,105 +20,96 @@
...
 }
 
-int32_t fix_ATan2(int32_t i, int32_t j)
+int32_t fix_ATan2(int32_t i, int32_t j, int32_t q)
 {
 	return double2fix(rad2turn(atan2(fix2double(i), fix2double(j))));
```
Comment (ISSOtm) [inline src/asm/fixpoint.c:74]:
```suggestion
	return double2fix(rad2turn(atan2(fix2double(i, q), fix2double(j, q))), q);
```

File: src/asm/fixpoint.c [hunk 9]
```diff
@@ -21,105 +20,96 @@
...
 }
 
-int32_t fix_Mul(int32_t i, int32_t j)
+int32_t fix_Mul(int32_t i, int32_t j, int32_t q)
 {
 	return double2fix(fix2double(i) * fix2double(j));
```
Comment (ISSOtm) [inline src/asm/fixpoint.c:79]:
```suggestion
	return double2fix(fix2double(i, q) * fix2double(j, q), q);
```

File: src/asm/fixpoint.c [hunk 10]
```diff
@@ -21,105 +20,96 @@
...
 }
 
-int32_t fix_Div(int32_t i, int32_t j)
+int32_t fix_Div(int32_t i, int32_t j, int32_t q)
 {
 	return double2fix(fix2double(i) / fix2double(j));
```
Comment (ISSOtm) [inline src/asm/fixpoint.c:84]:
```suggestion
	return double2fix(fix2double(i, q) / fix2double(j, q), q);
```

File: src/asm/fixpoint.c [hunk 11]
```diff
@@ -21,105 +20,96 @@
...
 }
 
-int32_t fix_Mod(int32_t i, int32_t j)
+int32_t fix_Mod(int32_t i, int32_t j, int32_t q)
 {
 	return double2fix(fmod(fix2double(i), fix2double(j)));
```
Comment (ISSOtm) [inline src/asm/fixpoint.c:89]:
```suggestion
	return double2fix(fmod(fix2double(i, q), fix2double(j, q)), q);
```

File: src/asm/fixpoint.c [hunk 12]
```diff
@@ -21,105 +20,96 @@
...
 }
 
-int32_t fix_Pow(int32_t i, int32_t j)
+int32_t fix_Pow(int32_t i, int32_t j, int32_t q)
 {
 	return double2fix(pow(fix2double(i), fix2double(j)));
```
Comment (ISSOtm) [inline src/asm/fixpoint.c:94]:
```suggestion
	return double2fix(pow(fix2double(i, q), fix2double(j, q)), q);
```

File: src/asm/fixpoint.c [hunk 13]
```diff
@@ -21,105 +20,96 @@
...
 }
 
-int32_t fix_Log(int32_t i, int32_t j)
+int32_t fix_Log(int32_t i, int32_t j, int32_t q)
 {
 	return double2fix(log(fix2double(i)) / log(fix2double(j)));
```
Comment (ISSOtm) [inline src/asm/fixpoint.c:99]:
```suggestion
	return double2fix(log(fix2double(i, q)) / log(fix2double(j, q)), q);
```

File: src/asm/parser.y [hunk 14]
```diff
@@ -1549,6 +1551,17 @@ const_no_str	: relocexpr_no_str { $$ = rpn_GetConstVal(&$1); }
 const_8bit	: reloc_8bit { $$ = rpn_GetConstVal(&$1); }
 ;
 
+opt_q_const_rparen	: T_RPAREN { $$ = fix_Precision(); }
+		| T_COMMA const T_RPAREN {
```
Comment (Rangi42) [inline src/asm/parser.y:1555]: See if this is unambiguous to parse:
```suggestion
opt_q_const	: %empty { $$ = fix_Precision(); }
		| T_COMMA const {
```

File: src/asm/fixpoint.c [hunk 15]
```diff
@@ -21,105 +20,96 @@
...
 }
 
-int32_t fix_Round(int32_t i)
+int32_t fix_Round(int32_t i, int32_t q)
 {
 	return double2fix(round(fix2double(i)));
```
Comment (ISSOtm) [inline src/asm/fixpoint.c:104]:
```suggestion
	return double2fix(round(fix2double(i, q)), q);
```

File: src/asm/fixpoint.c [hunk 16]
```diff
@@ -21,105 +20,96 @@
...
 }
 
-int32_t fix_Ceil(int32_t i)
+int32_t fix_Ceil(int32_t i, int32_t q)
 {
 	return double2fix(ceil(fix2double(i)));
```
Comment (ISSOtm) [inline src/asm/fixpoint.c:109]:
```suggestion
	return double2fix(ceil(fix2double(i, q)), q);
```

File: src/asm/fixpoint.c [hunk 17]
```diff
@@ -21,105 +20,96 @@
...
 }
 
-int32_t fix_Floor(int32_t i)
+int32_t fix_Floor(int32_t i, int32_t q)
 {
 	return double2fix(floor(fix2double(i)));
```
Comment (ISSOtm) [inline src/asm/fixpoint.c:114]:
```suggestion
	return double2fix(floor(fix2double(i, q)), q);
```

File: src/asm/parser.y [hunk 18]
```diff
@@ -1549,6 +1551,17 @@ const_no_str	: relocexpr_no_str { $$ = rpn_GetConstVal(&$1); }
...
+			if ($2 >= 1 && $2 <= 31) {
+				$$ = $2;
+			} else {
+				error("Fixed-point precision must be between 1 and 31\n");
+				$$ = fix_Precision();
+			}
+		}
+;
```
Comment (ISSOtm) [inline src/asm/parser.y:1563]:
```suggestion
opt_q_arg	: %empty { $$ = fix_Precision(); }
		| T_COMMA const {
			if ($2 >= 1 && $2 <= 31) {
				$$ = $2;
			} else {
				error("Fixed-point precision must be between 1 and 31\n");
				$$ = fix_Precision();
			}
		}
;
```

(...and associated changes.)

**Review by ISSOtm — CHANGES_REQUESTED** [review]

**Review by ISSOtm — APPROVED** [review]

**Pull request merged**